jormungandr/src/network/mod.rs

```diff
@@ -360,61 +337,92 @@ async fn handle_network_input(
+                if f(peer.addr, arg.clone()).await.is_err() {
+                    res.push(peer);
+                }
+            }
+            Ok(res)
+        }
+
         let prop_state = state.clone();
-        let propagate_res = match &msg {
+        let unreached_nodes = match &msg {
             PropagateMsg::Block(header) => {
                 tracing::debug!(hash = %header.hash(), "block to propagate");
                 let header = header.encode();
-                let view = state
-                    .topology
-                    .view(poldercast::layer::Selection::Topic {
-                        topic: p2p::topic::BLOCKS,
-                    })
-                    .await;
-                prop_state
-                    .peers
-                    .propagate_block(
-                        view.peers
-                            .into_iter()
-                            .map(|profile| (profile.address(), profile.id()))
-                            .collect(),
-                        header,
-                    )
-                    .await
+                propagate(
+                    |addr, header| prop_state.peers.propagate_block(addr, header),
+                    Selection::Topic {
+                        topic: crate::topology::topic::BLOCKS,
+                    },
+                    header,
+                    &mut channels.topology_box,
+                )
+                .await?
             }
             PropagateMsg::Fragment(fragment) => {
                 tracing::debug!(hash = %fragment.hash(), "fragment to propagate");
                 let fragment = fragment.encode();
-                let view = state
-                    .topology
-                    .view(poldercast::layer::Selection::Topic {
-                        topic: p2p::topic::MESSAGES,
-                    })
-                    .await;
-
-                prop_state
+                propagate(
+                    |addr, fragment| prop_state.peers.propagate_fragment(addr, fragment),
+                    Selection::Topic {
+                        topic: crate::topology::topic::MESSAGES,
+                    },
+                    fragment,
+                    &mut channels.topology_box,
+                )
+                .await?
+            }
+            PropagateMsg::Gossip(peer, gossips) => {
+                tracing::debug!("gossip to propagate");
+                let gossip = gossips.encode();
+                match prop_state
                     .peers
-                    .propagate_fragment(
-                        view.peers
-                            .into_iter()
-                            .map(|profile| (profile.address(), profile.id()))
-                            .collect(),
-                        fragment,
-                    )
+                    .propagate_gossip_to(peer.addr, gossip)
                     .await
+                {
+                    Err(_) => vec![peer.clone()],
```

Comment:
Maybe log the error here?

Reply:
I think it will be logged by `propagate_gossip` already, also because the `Err` variant contains no information about the error, just the address of the peer.